lib/sanity/attributable.rb

```diff
@@ -22,33 +29,46 @@ def attributes
         @attributes ||= []
       end
 
-      def default_attributes
-        @defaults ||= {}
+      def aliased_attributes
+        @aliased_attributes ||= {}
       end
 
       private
 
-      def attribute(name, default: nil)
-        attributes << name
-        default_attributes.merge!("#{name}": default)
+      def attribute(name, default: nil, as: nil, type: nil)
+        attributes.push(name)
+
+        self.define_method("#{name}=") do |value|
+          @attributes[name] =
+            case type
+            when :boolean
+              value.to_s == 'true'
+            when :float
+              value.to_f
+            when :integer
+              value.to_i
+            when :time
+              Time.parse(value.to_s)
+            else
+              value
+            end
+        end
+        self.define_method("#{name}") { @attributes[name] }
+
+        if as
+          aliased_attributes[as] = name
+          self.alias_method("#{as}=", "#{name}=")
+          self.alias_method(as, name)
+        end
       end
+
     end
 
     attr_reader :attributes
 
-    def initialize(**args)
-      self.class.default_attributes.merge(args).then do |attrs|
-        attrs.each do |key, val|
-          define_singleton_method("#{key}=") do |val|
-            args[key] = val
-            attributes[key] = val
-          end
-
-          define_singleton_method(key) { args[key] }
-        end
-
-        instance_variable_set(:@attributes, attrs)
-      end
+    def initialize(**kwargs)
+      @attributes ||= {}
+      kwargs.each { |name, value| public_send("#{name}=", value) }
```

Comment:
By not using define_singleton_method you're also discounting inheritance of this module. Any overwriting method of the attribute will not respond to super correctly. If you're not clear on define_singleton_method and what it does lmk and I can say more..

Reply:
I'm not familiar with that method or how and when to use it. My main goal here was making it so you could overwrite the methods for any attribute setter or getter. Previously you couldn't because the methods were defined on initialization of each instance instead of when loading the class.